include/drivers/usb/udc.h

```diff
@@ -0,0 +1,676 @@
+	uint32_t zlp : 1;
+	/** Flag marks request buffer claimed by the controller (TBD) */
+	uint32_t claimed : 1;
+	/** Flag marks request buffer is queued (TBD) */
+	uint32_t queued : 1;
+} __packed;
+
+/**
+ * @typedef udc_event_cb_t
+ * @brief Callback to submit UDC event to higher layer.
+ *
+ * At the higher level, the event is to be inserted into a message queue.
+ * (TBD) Maybe it is better to provide a pointer to k_msgq passed during initialization.
+ *
+ * @param[in] dev      Pointer to device struct of the driver instance
+ * @param[in] event    Point to event structure
+ *
+ * @return 0 on success, all other values should be treated as error.
+ */
+typedef int (*udc_event_cb_t)(const struct device *dev,
+			      struct udc_event *event);
+
+typedef int (*udc_lock_t)(const struct device *dev);
+
+typedef int (*udc_unlock_t)(const struct device *dev);
+
+typedef int (*udc_init_t)(const struct device *dev);
+
+typedef int (*udc_enable_t)(const struct device *dev);
+
+typedef int (*udc_disable_t)(const struct device *dev);
+
+typedef int (*udc_shutdown_t)(const struct device *dev);
+
+typedef int (*udc_set_address_t)(const struct device *dev,
+				 const uint8_t addr);
+
+typedef int (*udc_host_wakeup_t)(const struct device *dev);
+
+typedef int (*udc_ep_assign_t)(const struct device *dev,
+			       struct udc_ep_config *cfg);
+
+typedef int (*udc_ep_enable_t)(const struct device *dev,
+			       struct udc_ep_config *cfg);
+
+typedef int (*udc_ep_disable_t)(const struct device *dev,
+				struct udc_ep_config *cfg);
+
+typedef int (*udc_ep_unassign_t)(const struct device *dev,
+				 struct udc_ep_config *cfg);
+
+typedef int (*udc_ep_set_halt_t)(const struct device *dev,
+				 struct udc_ep_config *cfg);
+
+typedef int (*udc_ep_clear_halt_t)(const struct device *dev,
+				   struct udc_ep_config *cfg);
+
+typedef int (*udc_ep_flush_t)(const struct device *dev,
+			      struct udc_ep_config *cfg);
+
+typedef int (*udc_ep_buf_free_t)(const struct device *dev,
+				 struct net_buf *buf);
+
+typedef int (*udc_ep_buf_alloc_t)(const struct device *dev,
+				  struct net_buf *buf);
+
+typedef int (*udc_ep_queue_t)(const struct device *dev,
+			      struct udc_ep_config *cfg,
+			      struct net_buf *buf);
+
+typedef int (*udc_ep_dequeue_t)(const struct device *dev,
+				struct udc_ep_config *cfg);
+
+/**
+ * @brief UDC driver API
+ * This is the mandatory API any USB device controller driver needs to expose.
+ */
+struct udc_api {
+	udc_lock_t lock;
+	udc_unlock_t unlock;
+	udc_init_t init;
+	udc_enable_t enable;
+	udc_disable_t disable;
+	udc_shutdown_t shutdown;
+	udc_set_address_t set_address;
+	udc_host_wakeup_t host_wakeup;
+	udc_ep_assign_t ep_assign;
+	udc_ep_enable_t ep_enable;
+	udc_ep_disable_t ep_disable;
+	udc_ep_unassign_t ep_unassign;
+	udc_ep_set_halt_t ep_set_halt;
+	udc_ep_clear_halt_t ep_clear_halt;
+	udc_ep_flush_t ep_flush;
+	udc_ep_buf_free_t ep_free_buf;
+	udc_ep_buf_alloc_t ep_alloc_buf;
```

Comment:
Shouldn't this be mentioned that these two functions may have no implementation?

Reply:
That pars is kind of WiP, I will update that later (TODO)